src/relax/transform/lambda_lift.cc

```diff
@@ -336,176 +266,235 @@ class LambdaLifter : public ExprMutator {
+      lifted_func = GetRef<Function>(func_node);
+    } else {
+      lifted_func =
+          Function(lifted_func_params, body, ret_struct_info, func_node->is_pure, func_node->attrs);
+    }
+
+    for (Var param : lifted_func->params) {
+      CHECK(param->checked_type_.defined())
+          << "relax.Function requires all parameters to contain checked_type_.  "
+          << "However, parameter " << param << " with struct info " << param->struct_info_
+          << " has no checked type";
     }
 
     ICHECK(lifted_func.defined());
 
+    if (is_closure || IsClosure(lifted_func)) {
+      closures_.insert(gvar_lifted_func);
+    }
+
     // Add the lifted function to the module.
-    global->struct_info_ = GetStructInfo(lifted_func);
-    global->checked_type_ = lifted_func->checked_type_;
-    builder_->UpdateFunction(global, lifted_func);
+    lifted_func = CopyWithNewVars(lifted_func);
+    gvar_lifted_func->struct_info_ = GetStructInfo(lifted_func);
+    gvar_lifted_func->checked_type_ = lifted_func->checked_type_;
 
-    if (!is_closure) {
-      return std::move(global);
-    } else {
+    builder_->UpdateFunction(gvar_lifted_func, lifted_func);
+
+    Expr callable_value = gvar_lifted_func;
+    if (is_closure) {
       // If we need to allocate a closure,
       // we pass the variables in its environment here.
-      Array<Expr> fvs;
-      for (auto fv : captured_vars) {
-        fvs.push_back(fv);
-      }
+      Tuple arg_tuple(captured_vars.Map([](Var var) -> Expr { return var; }));
       // Call make_closure intrinsic
-      return Call(make_closure_op_, {global, Tuple(fvs)}, {}, {});
+      callable_value = Call(make_closure_op_, {gvar_lifted_func, arg_tuple}, {}, {});
     }
+
+    return callable_value;
   }
 
-  bool HasClosure(const Var& var) {
-    auto val = builder_->LookupBinding(var);
-    if (const auto* value = val.as<GlobalVarNode>()) {
-      IRModule ctx_mod = builder_->GetContextIRModule();
-      ICHECK(ctx_mod->functions.size() > 0);
-      BaseFunc func = ctx_mod->Lookup(GetRef<GlobalVar>(value));
-      if (const auto* func_node = func.as<FunctionNode>()) {
-        if (const auto* call_node = func_node->body.as<CallNode>()) {
-          if (call_node->op == make_closure_op_) {
-            return true;
-          }
-        } else if (const auto* seq_expr_node = func_node->body.as<SeqExprNode>()) {
-          // the return var points to a make_closure intrinsic
-          if (const auto* var = seq_expr_node->body.as<VarNode>()) {
-            return HasClosure(GetRef<Var>(var));
+  Expr VisitExpr_(const CallNode* call_node) final {
+    auto call = GetRef<Call>(call_node);
+
+    auto orig_sinfo = Downcast<StructInfo>(call->struct_info_);
+
+    if (auto opt_var = call->op.as<Var>()) {
+      auto var = opt_var.value();
+
+      // Call "relax.invoke_closure" to invoke closure
+
+      if (bool is_closure = IsClosure(var);
+          is_closure && builder_->LookupBinding(var).as<CallNode>()) {
```

Comment:
Is there a reason to write it this way instead of `IsClosure(var) && ...`?

Reply:
Nope.  A previous implementation of `IsClosure` returned an `Optional<_>`, and this stuck around.  Updated.